src/object_store/src/object/mod.rs

```diff
@@ -542,94 +537,117 @@ impl<OS: ObjectStore> MonitoredObjectStore<OS> {
             .observe(obj.len() as f64);
         let _timer = self
             .object_store_metrics
             .operation_latency
-            .with_label_values(&[self.media_type(), operation_type])
+            .with_label_values(&[self.media_type(), operation_type_str])
             .start_timer();
-        let future = async {
+
+        let builder = || async {
             self.inner
-                .upload(path, obj)
-                .verbose_instrument_await("object_store_upload")
+                .upload(path, obj.clone())
+                .verbose_instrument_await(operation_type_str)
                 .await
         };
-        let res = match self.upload_timeout.as_ref() {
-            None => future.await,
-            Some(timeout) => tokio::time::timeout(*timeout, future)
-                .await
-                .unwrap_or_else(|_| Err(ObjectError::internal("upload timeout"))),
-        };
 
-        try_update_failure_metric(&self.object_store_metrics, &res, operation_type);
+        let retry_condition = RetryCondition::new(
+            self.config.retry.upload_timeout_ms,
+            operation_type,
+            self.object_store_metrics.clone(),
+        );
+
+        let res = retry_request(builder, &self.config, operation_type, retry_condition).await;
+
+        try_update_failure_metric(&self.object_store_metrics, &res, operation_type_str);
         res
     }
 
     pub async fn streaming_upload(&self, path: &str) -> ObjectResult<MonitoredStreamingUploader> {
-        let operation_type = "streaming_upload_start";
+        let operation_type = OperationType::StreamingUploadInit;
+        let operation_type_str = operation_type.as_str();
         let media_type = self.media_type();
         let _timer = self
             .object_store_metrics
             .operation_latency
-            .with_label_values(&[media_type, operation_type])
+            .with_label_values(&[media_type, operation_type_str])
             .start_timer();
-        let future = async {
+
+        let builder = || async {
             self.inner
                 .streaming_upload(path)
-                .verbose_instrument_await("object_store_streaming_upload")
+                .verbose_instrument_await(operation_type_str)
                 .await
         };
-        let res = match self.streaming_upload_timeout.as_ref() {
-            None => future.await,
-            Some(timeout) => tokio::time::timeout(*timeout, future)
-                .await
-                .unwrap_or_else(|_| Err(ObjectError::internal("streaming_upload init timeout"))),
-        };
 
-        try_update_failure_metric(&self.object_store_metrics, &res, operation_type);
+        let retry_condition = RetryCondition::new(
+            self.config.retry.streaming_upload_timeout_ms,
+            operation_type,
+            self.object_store_metrics.clone(),
+        );
+
+        let res = retry_request(builder, &self.config, operation_type, retry_condition).await;
```

Comment:
Do we need a timeout for `streaming_upload` call? If yes, will it be clearer to move the init timeout logic under the specific streaming uploader implementaion given that the upload attmpt timeout is done inside the implementation. In this case, all streaming uploader related timeout is maintained inside the specific implementaion. Note that init timeout is unneccessary for `streaming_upload` under aws sdk.